### ISSUE TITLE
Linkerd-proxy sidecar containers Readiness probe fails and pod remains in NotReady state

### ISSUE THREAD

**Issue by kteague-tasktop** [linkerd/linkerd2#6255]
### What is the issue?

When a linkerd-proxy container is started, it never passes the Readiness probe and the pod remains in NotReady state.

(this is on pods within the linkerd control plane - we haven't started to extensively test using Linkerd with our app pods)

### How can it be reproduced?

When cluster nodes are terminated or rebooted, this condition can sometimes appear.

### Logs, error output, etc

Linkerd-proxy logs in DEBUG mode:

```
export POD=linkerd-tap-6ccd946964-2q7bh
kubectl -n linkerd logs ${POD} linkerd-proxy
time="2021-06-11T19:44:07Z" level=info msg="running version stable-2.9.4"
{"timestamp":"[   184.717153s]","level":"INFO","fields":{"message":"set new log level","level":"linkerd=debug"},"target":"linkerd2_tracing::level","threadId":"ThreadId(2)"}
{"timestamp":"[   184.736309s]","level":"DEBUG","fields":{"message":"Connection closed"},"target":"linkerd2_app_core::serve","threadId":"ThreadId(2)"}
{"timestamp":"[   184.955782s]","level":"DEBUG","fields":{"peer.identity":"Some(\"linkerd-identity.linkerd.serviceaccount.identity.linkerd.cluster.local\")"},"target":"linkerd2_proxy_transport::tls::client","threadId":"ThreadId(1)"}
{"timestamp":"[   184.955827s]","level":"DEBUG","fields":{"message":"Connecting","peer.addr":"10.137.11.19:8080"},"target":"linkerd2_proxy_transport::connect","threadId":"ThreadId(1)"}
{"timestamp":"[   184.978160s]","level":"DEBUG","fields":{"message":"Connected","local.addr":"10.137.76.44:39236","keepalive":"Some(10s)"},"target":"linkerd2_proxy_transport::connect","span":{"name":"h2"},"spans":[{"name":"h2"}],"threadId":"ThreadId(1)"}
{"timestamp":"[   184.979071s]","level":"DEBUG","fields":{"message":"Failed to connect","error":"received corrupt message"},"target":"linkerd2_reconnect::service","threadId":"ThreadId(1)"}
{"timestamp":"[   184.979592s]","level":"DEBUG","fields":{"message":"Recovering"},"target":"linkerd2_reconnect::service","threadId":"ThreadId(1)"}
{"timestamp":"[   184.995583s]","level":"DEBUG","fields":{"peer.identity":"Some(\"linkerd-identity.linkerd.serviceaccount.identity.linkerd.cluster.local\")"},"target":"linkerd2_proxy_transport::tls::client","threadId":"ThreadId(1)"}
{"timestamp":"[   184.996583s]","level":"DEBUG","fields":{"message":"Connecting","peer.addr":"10.137.89.182:8080"},"target":"linkerd2_proxy_transport::connect","threadId":"ThreadId(1)"}
{"timestamp":"[   184.997928s]","level":"DEBUG","fields":{"message":"Connected","local.addr":"10.137.76.44:34816","keepalive":"Some(10s)"},"target":"linkerd2_proxy_transport::connect","span":{"name":"h2"},"spans":[{"name":"h2"}],"threadId":"ThreadId(1)"}
{"timestamp":"[   184.998879s]","level":"DEBUG","fields":{"message":"Failed to connect","error":"received corrupt message"},"target":"linkerd2_reconnect::service","threadId":"ThreadId(1)"}
{"timestamp":"[   184.999006s]","level":"DEBUG","fields":{"message":"Recovering"},"target":"linkerd2_reconnect::service","threadId":"ThreadId(1)"}
{"timestamp":"[   185.165370s]","level":"DEBUG","fields":{"peer.identity":"Some(\"linkerd-identity.linkerd.serviceaccount.identity.linkerd.cluster.local\")"},"target":"linkerd2_proxy_transport::tls::client","threadId":"ThreadId(1)"}
{"timestamp":"[   185.165403s]","level":"DEBUG","fields":{"message":"Connecting","peer.addr":"10.137.68.246:8080"},"target":"linkerd2_proxy_transport::connect","threadId":"ThreadId(1)"}
{"timestamp":"[   185.165868s]","level":"DEBUG","fields":{"message":"Connected","local.addr":"10.137.76.44:55332","keepalive":"Some(10s)"},"target":"linkerd2_proxy_transport::connect","span":{"name":"h2"},"spans":[{"name":"h2"}],"threadId":"ThreadId(1)"}
{"timestamp":"[   185.166257s]","level":"DEBUG","fields":{"message":"Failed to connect","error":"received corrupt message"},"target":"linkerd2_reconnect::service","threadId":"ThreadId(1)"}
{"timestamp":"[   185.166280s]","level":"DEBUG","fields":{"message":"Recovering"},"target":"linkerd2_reconnect::service","threadId":"ThreadId(1)"}
{"timestamp":"[   185.480723s]","level":"DEBUG","fields":{"peer.identity":"Some(\"linkerd-identity.linkerd.serviceaccount.identity.linkerd.cluster.local\")"},"target":"linkerd2_proxy_transport::tls::client","threadId":"ThreadId(1)"}
{"timestamp":"[   185.480756s]","level":"DEBUG","fields":{"message":"Connecting","peer.addr":"10.137.11.19:8080"},"target":"linkerd2_proxy_transport::connect","threadId":"ThreadId(1)"}
{"timestamp":"[   185.481618s]","level":"DEBUG","fields":{"message":"Connected","local.addr":"10.137.76.44:39262","keepalive":"Some(10s)"},"target":"linkerd2_proxy_transport::connect","span":{"name":"h2"},"spans":[{"name":"h2"}],"threadId":"ThreadId(1)"}
{"timestamp":"[   185.482430s]","level":"DEBUG","fields":{"message":"Failed to connect","error":"received corrupt message"},"target":"linkerd2_reconnect::service","threadId":"ThreadId(1)"}
{"timestamp":"[   185.482451s]","level":"DEBUG","fields":{"message":"Recovering"},"target":"linkerd2_reconnect::service","threadId":"ThreadId(1)"}
{"timestamp":"[   185.499566s]","level":"DEBUG","fields":{"peer.identity":"Some(\"linkerd-identity.linkerd.serviceaccount.identity.linkerd.cluster.local\")"},"target":"linkerd2_proxy_transport::tls::client","threadId":"ThreadId(1)"}
{"timestamp":"[   185.499599s]","level":"DEBUG","fields":{"message":"Connecting","peer.addr":"10.137.89.182:8080"},"target":"linkerd2_proxy_transport::connect","threadId":"ThreadId(1)"}
{"timestamp":"[   185.500572s]","level":"DEBUG","fields":{"message":"Connected","local.addr":"10.137.76.44:34840","keepalive":"Some(10s)"},"target":"linkerd2_proxy_transport::connect","span":{"name":"h2"},"spans":[{"name":"h2"}],"threadId":"ThreadId(1)"}
{"timestamp":"[   185.501257s]","level":"DEBUG","fields":{"message":"Failed to connect","error":"received corrupt message"},"target":"linkerd2_reconnect::service","threadId":"ThreadId(1)"}
{"timestamp":"[   185.501280s]","level":"DEBUG","fields":{"message":"Recovering"},"target":"linkerd2_reconnect::service","threadId":"ThreadId(1)"}
{"timestamp":"[   185.666509s]","level":"DEBUG","fields":{"peer.identity":"Some(\"linkerd-identity.linkerd.serviceaccount.identity.linkerd.cluster.local\")"},"target":"linkerd2_proxy_transport::tls::client","threadId":"ThreadId(1)"}
{"timestamp":"[   185.666543s]","level":"DEBUG","fields":{"message":"Connecting","peer.addr":"10.137.68.246:8080"},"target":"linkerd2_proxy_transport::connect","threadId":"ThreadId(1)"}
```

#### `linkerd check` output

```
kubernetes-api
--------------
√ can initialize the client
√ can query the Kubernetes API

kubernetes-version
------------------
√ is running the minimum Kubernetes API version
√ is running the minimum kubectl version

linkerd-existence
-----------------
√ 'linkerd-config' config map exists
√ heartbeat ServiceAccount exist
√ control plane replica sets are ready
√ no unschedulable pods
√ controller pod is running

linkerd-config
--------------
√ control plane Namespace exists
√ control plane ClusterRoles exist
√ control plane ClusterRoleBindings exist
√ control plane ServiceAccounts exist
√ control plane CustomResourceDefinitions exist
√ control plane MutatingWebhookConfigurations exist
√ control plane ValidatingWebhookConfigurations exist
√ control plane PodSecurityPolicies exist

linkerd-cni-plugin
------------------
√ cni plugin ConfigMap exists
√ cni plugin PodSecurityPolicy exists
√ cni plugin ClusterRole exists
√ cni plugin ClusterRoleBinding exists
√ cni plugin Role exists
√ cni plugin RoleBinding exists
√ cni plugin ServiceAccount exists
√ cni plugin DaemonSet exists
√ cni plugin pod is running on all nodes

linkerd-identity
----------------
√ certificate config is valid
√ trust anchors are using supported crypto algorithm
√ trust anchors are within their validity period
√ trust anchors are valid for at least 60 days
√ issuer cert is using supported crypto algorithm
√ issuer cert is within its validity period
√ issuer cert is valid for at least 60 days
√ issuer cert is issued by the trust anchor

linkerd-webhooks-and-apisvc-tls
-------------------------------
√ proxy-injector webhook has valid cert
√ proxy-injector cert is valid for at least 60 days
√ sp-validator webhook has valid cert
√ sp-validator cert is valid for at least 60 days

linkerd-identity-data-plane
---------------------------
√ data plane proxies certificate match CA

linkerd-api
-----------
\ pod/linkerd-controller-69cd4489b9-ghsl2 container linkerd-proxy is not ready
```

### Environment

- Kubernetes Version: v1.18.9-eks-d1db3c
- Cluster Environment: AWS EKS on EC2 ASG
- Host OS: Ubuntu 20.04.2 LTS
- Linkerd version: 2.9.4 (originally seen on 2.9.2 but updated to 2.9.4 and still persists)

### Possible solution

Sometimes when this error happens, doing a rollout restart on `linkerd-identity` will allow a rollout restart to happen on the other control plane pods with stuck linkerd-proxy pods. Other times, restarting linkerd-identity does not change anything and the problem persists.

### Additional context

Our organization (Tasktop Technologies) has been setting up Linkerd as our service mesh, we have been running and testing it in a staging environment, but would like to use this in a production environment in about one month.

We have tested and seen this problem with both the normal LInkerd2 helm chart and with the HA helm chart.

**Comment by adleong** [comment]
Based on the logs you shared, it seems like the controller's linkerd-proxy is not able to establish a successful TLS connection to the identity controller.  Are you able to reproduce this error with a totally fresh install of Linkerd?  Or does it only occur after upgrading?  If so, can you share the exact steps (commands) used to upgrade Linkerd?

**Comment by kteague-tasktop** [comment]
I've just made a fresh install of Linkerd 2.9.4 in a new EKS Cluster. The problem appeared right away this time, with only linkerd-identity pods healthy and all other control plane pods having a NotReady linkerd-proxy.

**Comment by kteague-tasktop** [comment]
After doing a rolling restart to linkerd-identity on the fresh cluster, then linkerd-proxy was healthy. The linkerd-proxy pod for a previously stuck pod is logging what appears normal(?):

```
{"timestamp":"[  2976.866987s]","level":"DEBUG","fields":{"message":"Peeked bytes from TCP stream","sz":117},"target":"linkerd2_proxy_transport::tls::accept","span":{"peer.addr":"10.136.2.38:46084","target.addr":"10.136.2.13:4191","name":"accept"},"spans":[{"peer.addr":"10.136.2.38:46084","target.addr":"10.136.2.13:4191","name":"accept"}],"threadId":"ThreadId(2)"}
{"timestamp":"[  2976.867162s]","level":"DEBUG","fields":{"message":"Connection closed"},"target":"linkerd2_app_core::serve","span":{"peer.addr":"10.136.2.38:46084","target.addr":"10.136.2.13:4191","name":"accept"},"spans":[{"peer.addr":"10.136.2.38:46084","target.addr":"10.136.2.13:4191","name":"accept"}],"threadId":"ThreadId(2)"}
```

**Comment by kleimkuhler** [comment]
@kteague-tasktop So after restarting linkerd-identity, you are seeing the proxies for previously stuck pods starting up?

If you see this happen again, it'd be helpful if you can share the logs for linkerd-identity's proxy as well.

**Comment by kteague-tasktop** [comment]
We saw this again and spent some time debugging it - the messages coming from linkerd-identity appeared "bad" but we weren't able to determine the root cause. The fix was to remove the helm charts for both linkerd and linkerd-cni and re-install them and the service mesh came up functional (and only upon removal/re-install of linkerd-cni ... just re-installing linkerd wouldn't fix it this time).

However, we've just updated to 2.10.2 (and our scripting our staging nodes to run through a series of continual reboots as this problem always appears after a node reboot and pods are evicted and rescheduled) so we'll see if fixes in the latest release address this problem.

**Comment by kleimkuhler** [comment]
@kteague-tasktop Okay thanks for the update let us know if it seems to be fixed by the last stable and we'll go from there.

**Comment by olix0r** [comment]
Closing this for now, but please let us know if you hit anything like this again!

**Comment by kteague-tasktop** [comment]
We continued to see the problem on linkerd 2.10.2.

Howevrer, after seeing the issue, "linkerd + linkerd-cni does not recover when a node restarts": https://github.com/linkerd/linkerd2/issues/4789 - I looked into our k8s environments and we were running both aws-vpc-cni (for core CNI) and aws-calico-cni (for ACL like functions). We removed linkerd-cni and changed to linkerd-init InitContainers. Our stress test node rebooting has been running without errors since then.

There are various issues (in linkerd and aws-vpc-cni) about running linkerd-cni with other CNIs also in place. It would be good to have a note on the Linkerd CNI Plugin page to be cautious about running linkerd-cni with other CNIs and the potential for the CNIs to get out-of-sync or node reboot race conditions to appear.
